src/Examine.Core/ValueSet.cs

```diff
@@ -4,141 +4,317 @@
+        /// <param name="itemType">The index item type. Used to type the item in the index (in Umbraco terms this would be the document type alias).</param>
+        /// <param name="values">The values to be indexed.</param>
+        public ValueSet(string id, string category, string itemType, IDictionary<string, IEnumerable<object>> values)
+            : this(id, category, itemType, Yield(values))
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ValueSet" /> class.
+        /// </summary>
+        /// <param name="valueSet">The value set.</param>
+        public ValueSet(ValueSet valueSet)
+            : this(valueSet.Id, valueSet.Category, valueSet.ItemType, Yield(valueSet.Values))
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ValueSet" /> class.
+        /// </summary>
+        /// <param name="id">The identifier of the item to be indexed.</param>
+        /// <param name="category">The index category. Used to categorize the item in the index (in Umbraco terms this would be content, media or member).</param>
+        /// <param name="itemType">The index item type. Used to type the item in the index (in Umbraco terms this would be the document type alias).</param>
+        /// <param name="values">The values to be indexed.</param>
+        /// <remarks>
+        /// This constructor is not public, because we want to ensure values is not null and contains unique keys.
+        /// </remarks>
+        private ValueSet(string id, string category, string itemType, IEnumerable<KeyValuePair<string, IEnumerable<object>>> values)
         {
+            Id = id;
+            Category = category;
+            ItemType = itemType;
+            Values = values.ToDictionary<KeyValuePair<string, IEnumerable<object>>, string, IList<object>>(x => x.Key, x => x.Value.ToList());
         }
 
         /// <summary>
-        /// Primary constructor
+        /// Adds the value.
         /// </summary>
-        /// <param name="id"></param>
-        /// <param name="itemType">
-        /// The item's node type (in umbraco terms this would be the doc type alias)</param>
-        /// <param name="category">
-        /// Used to categorize the item in the index (in umbraco terms this would be content vs media)
-        /// </param>
-        /// <param name="values"></param>
-        public ValueSet(string id, string category, string itemType, IDictionary<string, IEnumerable<object>> values)
-            : this(id, category, itemType, values.ToDictionary(x => x.Key, x => (IReadOnlyList<object>)x.Value.ToList()))
+        /// <param name="key">The key.</param>
+        /// <param name="value">The value.</param>
+        /// <returns>
+        /// The number of values stored for the specified key.
+        /// </returns>
+        public int AddValue(string key, object value)
```

Comment:
I've renamed the `Add()`, `Set()` and `TryAdd()` methods to a more explicit `AddValues()`, `SetValues()` and `TryAddValues()`, which also aligns better with the existing `GetValues()` one.